config/jobs/cert-manager/cert-manager-presubmits.yaml

```diff
@@ -518,16 +518,19 @@ presubmits:
           value: "1"
 
     # This is the default e2e test for all PRs.
-  - name: pull-cert-manager-e2e-v1-23
+  - &default-e2e
+    name: pull-cert-manager-e2e-v1-23
     context: pull-cert-manager-e2e-v1-23
-    always_run: true
-    optional: false
+    always_run: false
```

Comment:
I think that we want to always run this test against all PRs against current master. This change will force the PR authors to manually run a command to run the test I think?

Reply:
Yes, that's my mistake, I removed `always_run: true` but should not have.

Reply:
Perhaps you want to create a second job where `always_run: false` and branches are set to `release-*` and keep this one restricted to just master and release-1.7?

Reply:
Unfortunately, the `always_run` flag doesn't look at the `branches`. If I had two separate jobs


```yaml
  - always_run: true
    optional: false
    branches:
    - master
    - release-1.7
  - always_run: false
    optional: true
    skip_branches:
    - release-1.5
```

then the first job would still be triggered on release-1.5 due to `always_run`. I think we are misusing `always_run`. Shouldn't we rather use `optional: false` so that the job is run for the branches listed in `branches`?

Reply:
Do I understand correctly:

- `optional: true` means that this job must report a GitHub check ([docs](https://github.com/kubernetes/test-infra/blob/master/prow/jobs.md#requiring-jobs-for-auto-merge-through-tide)),
- `always_run: true` means that this job will be automatically run when a PR is opened.

I definitely think `always_run` is buggy since it doesn't account for `branches` nor `skip_branches`.

Reply:
To see if the `always_run: true` behavior was correct, I cloned the test-infra repo (where Prow is) and added two test cases:

```go
{
	name: "job with always_run: true on a skipped branch should not run",
	job: Presubmit{
		Brancher: Brancher{
			SkipBranches: []string{"skipped"},
		},
		AlwaysRun: true,
	},
	ref:         "skipped",
	expectedRun: false,
},
{
	name: "job with always_run: true on a branch not listed in 'branches' should not run",
	job: Presubmit{
		Brancher: Brancher{
			Branches: []string{"master"},
		},
		AlwaysRun: true,
	},
	ref:         "not-listed-in-the-branches-filter",
	expectedRun: false,
},
```

and the tests passed. So `always_run: true` is working as expected.

Reply:
> Unfortunately, the always_run flag doesn't look at the branches. If I had two separate jobs

I'm not sure I understand what you mean. I meant keeping the existing job as in and adding one more that has both `always_run: false` and `optional: true` as well as `release-*`. The first one should then work in the same way as now (i.e run only against PRs for `master` and `release-1.7` and the second one should work in the same way as other modified jobs in this PR (i.e run on a PR against any branch, but only if triggered). Or, which one of those would you expect to misbehave?

Reply:
I think I was confused by the fact that I noticed that `pull-cert-manager-e2e-v1-23` was showing as a mandatory check on my PR against `release-1.5`:

![Screenshot from 2022-01-26 15-18-40](https://user-images.githubusercontent.com/2195781/151179832-e8b463d5-9108-4bad-82f0-1ee45031c714.png)


I just realized that this mandatory check isn't related to `always_run: true` nor `optional: false`.

It is due to another Prow configuration, in [config.yaml](https://github.com/maelvls/testing/blob/785481625ab73f45d092f5a3bd345919f5ddd466/config/config.yaml#L52-L67):

```yaml
    jetstack:
      repos:
        cert-manager:
          protect: true
          required_status_checks:
            contexts:
            - dco
            - pull-cert-manager-bazel
            - pull-cert-manager-deps
            - pull-cert-manager-chart
            - pull-cert-manager-e2e-v1-23
```

I will do as you suggested.

Reply:
@irbekrm I ended up just adding `release-1.5` to the previous presubmits.